pkg/devicetemplateconverter/grpc.go

```diff
@@ -39,25 +44,31 @@ func (s *endDeviceTemplateConverterServer) ListFormats(ctx context.Context, _ *p
 }
 
 // Convert implements ttnpb.DeviceTemplateServiceServer.
-func (s *endDeviceTemplateConverterServer) Convert(req *ttnpb.ConvertEndDeviceTemplateRequest, res ttnpb.EndDeviceTemplateConverter_ConvertServer) error {
+func (s *endDeviceTemplateConverterServer) Convert(
+	req *ttnpb.ConvertEndDeviceTemplateRequest,
+	res ttnpb.EndDeviceTemplateConverter_ConvertServer,
+) error {
 	converter, ok := s.DTC.converters[req.FormatId]
 	if !ok {
 		return errNotFound.WithAttributes("id", req.FormatId)
 	}
 	ctx, cancel := errorcontext.New(res.Context())
+	ctx = devicetemplates.NewContextWithProfileIDs(ctx, req.GetEndDeviceVersionIds())
+	ctx = profilefetcher.NewContextWithFetcher(ctx, profilefetcher.NewTemplateFetcher(s.DTC.Component))
 	ch := make(chan *ttnpb.EndDeviceTemplate)
 	go func() {
 		if err := converter.Convert(ctx, bytes.NewReader(req.Data), ch); err != nil {
 			cancel(err)
 		}
 	}()
+
 	for {
 		select {
 		case <-ctx.Done():
 			return ctx.Err()
 		case tmpl, ok := <-ch:
 			if !ok {
-				return nil
+				return ctx.Err()
```

Comment:
Why is this?

This would be a race between `ch` close and `ctx` done, so it's probably negligble in practice. However, theoterically, if the producer is done and closes `ch`, I don't know why a `ctx` error would be relevant. Why would we timeout if we have everything?

Reply:
https://github.com/TheThingsNetwork/lorawan-stack/pull/5610#discussion_r949211645

The idea behind it was that even when there is an error the channel is closed and therefore it isn't guaranteed that the flow will be directed to:
```go
case <-ctx.Done():
	return ctx.Err()
```
So there could be cases where there is an error, the channel is closed by the converter when returning said error and the select would be somewhat ambiguous. Meaning that `ctx.Err` is defined but since the channel is closed it could enter the case statement below and not provide such error to the user.
```go
case tmpl, ok := <-ch:
	if !ok {
		return nil
	}
```

At least this was the explanation that made sense for the error that I encountered while testing. Considering that the `ctx.Error` is `nil` whenever it is not set I approached this in a way that I wouldn't change the code too much.